File: fub/lib/fub/dir_stack.ex
defmodule Fub.DirStack do
  require Logger
  defstruct [:stack, :position]

  def new() do
    %__MODULE__{stack: [], position: 0}
  end

  def push(state, entry) do
    %{state | position: 0, stack: [entry | state.stack]}
  end

  def back(%__MODULE__{stack: []}) do
    :empty
  end

  def back(%__MODULE__{position: position} = state, entry) do
    state =
      if position == 0 do
        state
        |> push(entry)
        |> dedup()
      else
        state
      end

    position = min(position + 1, length(state.stack) - 1)

    {
      Enum.at(state.stack, position),
      %{state | position: position}
    }
  end

  def forward(%__MODULE__{stack: []}) do
    :empty
  end

  def forward(%__MODULE__{stack: stack, position: position} = state) do
    position = max(position - 1, 0)

    {
      Enum.at(stack, position),
      %{state | position: position}
    }
  end

  defp dedup(%__MODULE__{} = state) do
    %{state | stack: Enum.dedup(state.stack)}
  end
end

File: rurt/src/dir_stack.rs
use std::collections::VecDeque;
use std::hash::Hash;

#[derive(Default)]
pub struct DirStack<T> {
    stack: VecDeque<T>,
    position: usize,
}

impl<T: Clone + Eq + Hash + std::fmt::Debug> DirStack<T> {
    pub fn push(&mut self, entry: T) {
        if self.stack.front() == Some(&entry) {
            return;
        }

        self.position = 0;
        self.stack.push_front(entry);
    }

    pub fn back(&mut self, entry: T) -> Option<T> {
        if self.stack.is_empty() {
            return None;
        }


        if self.position == 0 {
            self.stack.push_front(entry);
            self.dedup();
        }

        self.position = (self.position + 1).min(self.stack.len() - 1);
        self.stack.get(self.position).cloned()
    }

    pub fn forward(&mut self) -> Option<T> {
        if self.stack.is_empty() {
            return None;
        }

        self.position = self.position.saturating_sub(1);

        self.stack.get(self.position).cloned()
    }

    fn dedup(&mut self) {
        let mut seen = std::collections::HashSet::new();
        self.stack.retain(|entry| seen.insert(entry.clone()));
    }
}
